File: node/vigoss.js
const express = require('express');
const mongoose = require('mongoose');

mongoose.Promise = global.Promise;

const db = mongoose.connect('mongodb://localhost:27017/poetry');

const schema = mongoose.Schema({
    title: String,
    author: String,
    content: [String]
});

const Poem = mongoose.model('songPoetry', schema, 'songPoetry');
const PoemN = mongoose.model('northernSongPoem', schema, 'northernSongPoem');
const PoemS = mongoose.model('southernSongPoem', schema, 'southernSongPoem');

const app = express();

const poemTypes = { songPoetry: Poem, northernSongPoem: PoemN, southernSongPoem: PoemS };

app.get('/', (req, res) => {
    res.sendFile('/home/node/poetry.html');
});

app.get('/poetry/:poemType/author/:name', (req, res) => {
    const poemType = req.params.poemType;
    const author = req.params.name;
    poemTypes[poemType].find({ author }).then(docs => {
        res.send(JSON.stringify(docs));
    }, err => {
        console.log(err);
    });
});

app.get('/poetry/:poemType', (req, res) => {
    const poemType = req.params.poemType;
    poemTypes[poemType].aggregate({ $group: { _id:'$author', num: { $sum: 1 } } }, { $sort: { num: -1 } }).then(docs => {
        const poets = docs.map(doc => doc._id);
        res.send(JSON.stringify(poets));
    }, err => {
        console.log(err);
    });
})

app.listen(3000, () => {
    console.log('Running.');
});

File: python/vigoss.py
from flask import Flask, jsonify
from pymongo import MongoClient

app = Flask(__name__)
poetList = {}

@app.route('/')
def home():
    return app.send_static_file('poetry.html')

@app.route('/poetry/<poemType>')
def poets(poemType):
    if poemType in poetList:
        poets = poetList[poemType]
    else:
        collection = db[poemType]
        pipeline = [
            { '$group': { '_id': '$author', 'num': { '$sum': 1 } } },
            { '$sort': { 'num': -1 } }
        ]
        cursor = collection.aggregate(pipeline)
        poets = list(map(lambda x: x['_id'], cursor))
        poetList[poemType] = poets
    return jsonify(poets)

def changeObjectId(d):
    d['_id'] = d['content'][0]
    return d

@app.route('/poetry/<poemType>/author/<author>')
def peoms(poemType, author):
    collection = db[poemType]
    cursor = collection.find({ 'author': author })
    poems = list(map(changeObjectId, cursor))
    return jsonify(poems)

if __name__ == '__main__':
    client = MongoClient()
    db = client.poetry

    app.run(port=3000)